python/ray/dag/compiled_dag_node.py

```diff
@@ -1110,34 +1114,76 @@ def _get_or_compile(
 
         # Create executable tasks for each actor
         for actor_handle, tasks in self.actor_to_tasks.items():
-            executable_tasks = []
+            # The key is the channel that the actor is reading from,
+            # the value is the number of times this channel is read
+            # by the actor.
+            actor_input_channels = defaultdict(int)
+
+            # Step 1: populate actor_input_channels and perform some validation.
             for task in tasks:
-                resolved_args = []
                 has_at_least_one_channel_input = False
                 for arg in task.args:
                     if isinstance(arg, InputNode):
-                        input_adapter = DAGInputAdapter(None, self.dag_input_channel)
-                        resolved_args.append(input_adapter)
+                        actor_input_channels[self.dag_input_channel] += 1
                         has_at_least_one_channel_input = True
                     elif isinstance(arg, InputAttributeNode):
-                        input_adapter = DAGInputAdapter(arg, self.dag_input_channel)
-                        resolved_args.append(input_adapter)
+                        actor_input_channels[self.dag_input_channel] += 1
                         has_at_least_one_channel_input = True
                     elif isinstance(arg, DAGNode):  # Other DAGNodes
                         arg_idx = self.dag_node_to_idx[arg]
                         arg_channel = self.idx_to_task[arg_idx].output_channel
                         assert arg_channel is not None
-                        resolved_args.append(arg_channel)
+                        actor_input_channels[self.arg_channel] += 1
                         has_at_least_one_channel_input = True
-                    else:
-                        resolved_args.append(arg)
                 # TODO: Support no-input DAGs (use an empty object to signal).
                 if not has_at_least_one_channel_input:
                     raise ValueError(
                         "Compiled DAGs require each task to take a "
                         "ray.dag.InputNode or at least one other DAGNode as an "
                         "input"
                     )
+
+            # Step 2: create cached channels if needed
+
+            # The key is the original input channel, the value is the cached channel
+            # to use in task execution.
+            actor_cached_channels = {}
+            for channel, count in actor_input_channels.items():
+                if count > 1:
+                    actor_cached_channels[channel] = CachedChannel(count, channel)
```

Comment:
Later code would be simpler if we overwrite self.dag_input_channel / task.output_channels (or add an actor_cached_channels entry for every channel instead of just the ones with count > 1).